ui/helpers/taskcluster.js

```diff
@@ -1,48 +1,115 @@
-import { OIDCCredentialAgent, Queue } from 'taskcluster-client-web';
+import { Queue } from 'taskcluster-client-web';
+import debounce from 'lodash/debounce';
+import delay from 'lodash/delay';
+import moment from 'moment';
 
-import { loginRootUrl, getUserSessionUrl } from './url';
+import { clientId, redirectURI } from '../taskcluster-auth-callback/constants';
+
+import { createQueryParams } from './url';
+
+export const tcCredentialsMessage =
+  'Need to retrieve or renew Taskcluster credentials before action can be performed.';
 
 const taskcluster = (() => {
-  let credentialAgent = null;
+  let _rootUrl = 'https://firefox-ci-tc.services.mozilla.com';
+
+  // from the MDN crypto.getRandomValues doc
+  const secureRandom = () =>
+    window.crypto.getRandomValues(new Uint32Array(1))[0] / 4294967295;
 
-  // Create an OIDC credential agent if it doesn't exist.
-  const tcAgent = () => {
-    if (credentialAgent) {
-      return credentialAgent;
+  const generateNonce = () => {
+    let value = '';
+    const characters =
+      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+    for (let i = 0; i <= 20; i++) {
+      value += characters.charAt(
+        Math.floor(secureRandom() * characters.length),
+      );
     }
 
-    const userSession = localStorage.getItem('userSession');
-    const oidcProvider = 'mozilla-auth0';
+    localStorage.setItem('requestState', value);
+    return value;
+  };
+
+  const getAuthCode = (useExistingWindow = false) => {
+    const nonce = generateNonce();
+    // we're storing these for use in the TaskclusterCallback component (taskcluster-auth.html)
+    // since that's the only way for it to get access to them
+    localStorage.setItem('requestState', nonce);
+    localStorage.setItem('tcRootUrl', _rootUrl);
+
+    // the expires param is optional and if it's greater than the maxExpires value that's been
+    // registered with the client (set to 3 days) or is omitted, expiry will default to maxExpires
+    const params = {
+      client_id: clientId,
+      response_type: 'code',
+      redirect_uri: redirectURI,
+      scope: 'hooks:trigger-hook:*',
+      state: nonce,
+      expires: '5 minutes',
```

Comment:
We should probably remove this line before merging so that clients are configured to live for 3 days (maxExpires) :)

Reply:
Yes, I mentioned that to sheriffs already - it was just added for testing.